instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/mappers/attribute.rb

```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module ActiveJob
+      module Mappers
+        # Maps ActiveJob Attributes to Semantic Conventions
+        class Attribute
+          # Generates a set of attributes to add to a span using
+          # general and messaging semantic conventions as well as
+          # using `rails.active_job.*` namespace for custom attributes
+          #
+          # @param payload [Hash] of an ActiveSupport::Notifications payload
+          # @return [Hash<String, Object>] of semantic attributes
+          def call(payload)
+            job = payload.fetch(:job)
+
+            otel_attributes = {
+              'code.namespace' => job.class.name,
+              'messaging.system' => job.class.queue_adapter_name,
+              'messaging.destination' => job.queue_name,
+              'messaging.message.id' => job.job_id,
+              'rails.active_job.execution.counter' => job.executions.to_i,
+              'rails.active_job.provider_job_id' => job.provider_job_id.to_s,
+              'rails.active_job.priority' => job.priority, # this can be problematic. Programs may use invalid attributes for priority.
```

Comment:
Would adding to_s (e.g. job.priority.to_s) make sense to overcome invalid attributes for priority?

Reply:
`NilClass` will render an empty string. Do we want the key to always be present even if the value is `nil`?

Reply:
What values could appear here that would be invalid? Invalid by whose rules, semantic convention†, OTLP format for attributes, or something else?

† I haven't found any, I think because a "job" namespace has yet to be defined.

Reply:
From https://opentelemetry.io/docs/specs/otel/common/#attribute

> An Attribute is a key-value pair, which MUST have the following properties:
> 
> - The attribute key MUST be a non-null and non-empty string.
> - The attribute value is either:
>   - A primitive type: string, boolean, double precision floating point (IEEE 754-1985) or signed 64 bit integer.
>   - An array of primitive type values. The array MUST be homogeneous, i.e., it MUST NOT contain values of different types.

Developers may set any value to `priority`. I ran into an issue where some teams were setting the value to a symbol e.g. `:low`, `:high`, and then doing custom processing of the `priority` values instead of using what was built into ActiveJob itself. 

The `priority` value was then being checked by the SDK, which prohibits `Symbol` types and would result in `OpenTelemetry.handle_error` being called for an invalid parameter.

Reply:
> † I haven't found any, I think because a "job" namespace has yet to be defined.

Bike shedding time. I do not think there are special Job semantics that really differ from messaging semantics, specifically for ActiveJob anyway. 

When I think of "job" processing on the other hand I think of `CronJob` as a concrete example of a function that runs on an interval and may or may not produce asynchronous events for processing. The "producer" in this case is the schedular and the event only has a timestamp that executes the work and the `consumer` is the script that is being executed. The semantics for me don't seem to line up cleanly with messaging because there isn't a concept of a "destination" or a "message" to process really.      

ActiveJob on the other hand is an asynchronous event processing system, whose specification describes the use of a single publisher (producer) and worker (consumer) per event. Though you could probably write an adapter that fanouts the event to _multiple destinations_ and to me this OOTB functionality the same semantics as point-to-point messaging, that is one producer-consumer with a dedicated queue (inline, in memory, Redis, postgresql, etc...).

It's unclear to me if attributes like `priorty` are generic for messaging systems, so my first pass at this was to use framework specific semconv. 

I asked about whether or not upstream semconv for framework specific attributes was welcome... see this discussion: 

https://cloud-native.slack.com/archives/C041APFBYQP/p1695733952824799

TL;DR each language should maintain its own set of semconv for framework specific components. 

All that being said, I just took another look at the messaging spec[^1].

> All attributes that are specific for a messaging system SHOULD be populated in messaging.{system} namespace. Attributes that describe a message, a destination, a consumer, or a batch of messages SHOULD be populated under the corresponding namespace:

Now I am second guessing my decision to use a `rails.active_job` namespace in this PR. 

Based on spec guidance does it make sense to use `messaging.active_job.destination.*` or `messaging.active_job.consumer.*` attributes?

Should we bother including these attributes at all? 

Does anyone need them?

[^1]: https://opentelemetry.io/docs/specs/semconv/messaging/messaging-spans/#attributes-specific-to-certain-messaging-systems

Reply:
If we're using the messaging semconv, I think yes to `messaging.active_job.*`. I admit that feels weird from an "I'm a Rails dev wondering what's going on in my system" perspective, though.

Reply:
🤔 What do you think of shipping with `messaging.active_job.*` (no `rails.` prefix) so we've hewed as close to OTel conventions as possible? Then, if we get feedback that these are weird names from a Rails perspective on conventions, we can decide to add more attributes under different names later.

Reply:
> 🤔 What do you think of shipping with messaging.active_job.* (no rails. prefix) so we've hewed as close to OTel conventions as possible?
> Then, if we get feedback that these are weird names from a Rails perspective on conventions, we can decide to add more attributes under different names later.

Yes! I plan on fixing that, though I think I am going to omit some of these attributes like the `execution_count` and `scheduled_at` attributes because I am dubious about their usefulness as span attributes.

Reply:
@robbkidd done! https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/677/commits/c0655ebaf0eb120d78026ee43fef33b65333cfe7

Reply:
You're the best!